### ISSUE TITLE
Jdbi should support function arguments similar to consumer arguments

### ISSUE THREAD

**Issue by hgschmie** [jdbi/jdbi#2326]
We support

```java
@SqlQuery("...") 
@UseRowMapper(SomethingMapper.class)
void foo(Consumer<Something> somethingConsumer); 
```

and especially

```java
@SqlQuery("...") 
@UseRowMapper(SomethingMapper.class)
void foo(Consumer<Stream<Something>> somethingConsumer); 
```

for methods today. Jdbi should also support

```java
@SqlQuery("...") 
@UseRowMapper(SomethingMapper.class)
ResultThing foo(Function<Stream<Something>, ResultThing> somethingFunction); 
``` 
where the right side of the Function argument must match the return type of the method. This would make some of the current stream/iterator consumption code much simpler:

```java
List<Something> result = new ArrayList();
jdbi.withExtension(SomethingDao.class, dao -> dao.loadAllStuff(stream -> stream.forEach(result::add)));

interface SomethingDao {
    @SqlQuery("...")
    void loadAllStuff(Consumer<Stream<Something>> consumer);
}
```

would become

```java
List<Something> result = jdbi.withExtension(SomethingDao.class, 
dao -> dao.loadAllStuff(stream -> stream.collect(Collectors.toList())));

interface SomethingDao {
    @SqlQuery("...")
    List<Something> loadAllStuff(Function<Stream<Something>, List<Something>> function);
}
```

This requires some work in the ResultReturner where the consumer argument detection code is located. It should also enforce the `<S, T> T someMethod(Function<S, T> functionArgument)` pattern.

**Comment by hgschmie** [comment]
There is a lot of code in SqlCallHandler#createResultTransformer that already does most of the work here. This should be aligned to the code in ResultReturner that deals with all the special cases for the consumers.

Also the SqlCallHandler code should probably not do its own thing for functions and consumers by calling Call#invoke() but use the Call#invoke(Consumer<>) and Call#invoke(Function<>) methods directly

**Comment by hgschmie** [comment]
Shipped in 3.41.x
